### PR TITLE
fix: stop confirming outputs before parsing them

### LINKED ISSUE #68: Change outputs are confirmed without being received
## Description

When a transaction is divided in chunks and the last chunk only has change outputs, the change outputs on the last chunk are confirmed before being received by the device, which sends the signing to the next step.
When the client tries to send the next tx chunk it fails with `BAD_STATE` due to state protection.

## Reproducing

A transaction that mimics this scenario without custom tokens (to make this simpler) was made, divided in chunks and added APDU headers.

chunk 1

`e0060000ff010400058000002c8000011880000000000000000000000002058000002c8000011880000000000000000000000203058000002c8000011880000000000000000000000304058000002c800001188000000000000000000000040001000205d57eff854a615612b5d089f667a32161b347344dd62dd7315f68dd508b17e1ddb300006d39806ee6bb56cc06b50723be09506b11ec11dde388a720ef1f68a9b2e6950aa200000000133800001976a91483d3ee6df8bee68f8043d3d7f7feba807a683dfe88ac000018f100001976a9142d9d5cc2969d59836204376a5c4927f84ca4c7b288ac0000113c00001976a91407a6e2e3c254b330de8d1b73bb3418e0`

chunk 2

`e0060001469466b45388ac0000206800001976a914279116ab46ab6111641fe2ad24bd8bf5b9636ad588ac000023b200001976a914d035b2d6e8663bbaa6c03e69df6953d5be25575288ac`

## Solution

Stop confirming outputs when we reach the last output received in the latest APDU call.
This means we need to keep track of the number of parsed outputs and only confirm outputs until the parsed ones, anything more should throw a request for more data.

### LINKED ISSUE #69: Change output list should be verified to avoid breaking the UI
## Description

If the change list has duplicated outputs, meaning 2 entries saying that output index X is a change, we will break the output confirming UI.
For instance, if we have 3 outputs and 2 entries on the change list saying that output index 0 is change we will show to the user to confirm outputs:

- `Output 1/1`
- `Output 2/1`

This can be specially bad if we had 2 outputs, since it actually shows `Output 1/0`, it can also show `Output 1/-1` depending on how we manipulate this.

## Solution

When parsing the change list we need to check for duplicates and fail the transaction if any are found.
Since we do not have data on the transaction yet we can parse the change list and when we parse the number of outputs on the transaction we also need to check that the highest change index is on the transaction (since we do not have duplicates this also checks that the number of change outputs is lower than the number of outputs).

### PR DESCRIPTION
**Pull request by r4mmer**
## Acceptance criteria

Solve issue #68

### CODE REVIEW

File: src/ui/display.c [hunk 1]
```diff
@@ -259,6 +259,13 @@ bool skip_change_outputs() {
     inplace_selection_sort((size_t) G_context.tx_info.change_len, change_indices);
 
     for (uint8_t i = 0; i < G_context.tx_info.change_len; i++) {
+        // current_output is the number of parsed outputs
+        // If it is the same as confirmed_outputs we have confirmed all parsed outputs
+        if (G_context.tx_info.confirmed_outputs >= G_context.tx_info.current_output) {
+            // We have confirmed all parsed outputs, we need to request more before continuiing
+            if (check_output_index_state()) return true;
```
Comment (r4mmer) [inline src/ui/display.c:266]: We need to call `check_output_index_state` here because not doing so creates an edge case where all outputs are confirmed but we do not ask the user to confirm the transaction.

File: tests/app_client/cmd.py [hunk 2]
```diff
@@ -118,6 +119,7 @@ def sign_tx(
             if sw != 0x9000:
                 raise DeviceException(error_code=sw, ins=InsType.INS_SIGN_TX)
 
+        time.sleep(0.1)
```
Comment (r4mmer) [inline tests/app_client/cmd.py:122]: Some tests would fail because speculos automation would be slower than pytest and we would send data before the device is ready.

File: tests/test_sign_edge.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+from hathorlib.scripts import P2PKH
+from hathorlib.utils import get_address_from_public_key_hash, get_hash160
+
+from app_client.transaction import ChangeInfo, TxInput, TxOutput
+from utils import fake_path, fake_tx
+
+fake = Faker()
+
+
+def test_last_chunk_only_change(cmd, public_key_bytes):
```
Comment (r4mmer) [inline tests/test_sign_edge.py:11]: public_key_bytes is a list of public keys derived from the speculos key at `m/44'/280'/0'/0/index`
This is to create actual scripts, that can be validated as change outputs by the device.

File: tests/test_sign_edge.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    This test creates a transaction that will be sent on 2 chunks.
+    The last chunk will have 1 partial and 2 complete outputs, all 3 are change
+    The change list is [0, 2, 3, 4] because we need the last output of the
+    first chunk (output index 1) to be confirmed manually, this would trigger
+    the auto confirmation of the next chunk.
+    This is described on issue #68
+    https://github.com/HathorNetwork/hathor-ledger-app/issues/68
```
Comment (r4mmer) [inline tests/test_sign_edge.py:19]: This is the explanation of how to trigger the issue being solved here.
This test was created on dev so it always failed and then moved to this branch so I can check that it always works.
In doing so I also found the edge case that using `check_output_index_state` fixes.

The values, tx_ids and scripts are randomly generated with Faker but the serialized value will have the same length, so the test will be consistent on this case.

File: tests/test_sign_cmd.py [hunk 5]
```diff
@@ -55,13 +51,13 @@ def test_sign_tx_change_protocol_v1(cmd, public_key_bytes):
         )
         for x in range(5)
     ]
-    change_indices = [fake.pyint(0, 5) for x in range(5)]
+    change_indices = list(set([fake.pyint(0, 4) for x in range(5)]))
```
Comment (r4mmer) [inline tests/test_sign_cmd.py:54]: This is because the range was [0, 5] meaning we may add an output index not on the tx (the 5 when indexes go up to 4).

I also changed it to be a `list(set(...))` to avoid repeating values. This could be done with `fake.unique.pyint` but it would just generate a `[0, 1, 2, 3, 4]` array (in a random order). My approach actually changes the length of the array, having always at least 1 change and up to 5.

File: tests/test_sign_edge.py (hunk 3 above)
Comment (tuliomir) [reply tests/test_sign_edge.py:11]: question(non-blocking): Why not add this comment to the code directly as a docstring? It seems important for future maintenance.

**Review by tuliomir — APPROVED** [review]

File: tests/test_sign_edge.py (hunk 3 above)
Comment (r4mmer) [reply tests/test_sign_edge.py:11]: This is part of the pytest injection system, its configured on the conftest.py file and there we have an explanation for this.
Also, the first thing on the file is the speculos default seed for testing.

File: src/ui/display.c [hunk 6]
```diff
@@ -259,6 +259,13 @@ bool skip_change_outputs() {
     inplace_selection_sort((size_t) G_context.tx_info.change_len, change_indices);
 
     for (uint8_t i = 0; i < G_context.tx_info.change_len; i++) {
+        // current_output is the number of parsed outputs
+        // If it is the same as confirmed_outputs we have confirmed all parsed outputs
+        if (G_context.tx_info.confirmed_outputs >= G_context.tx_info.current_output) {
+            // We have confirmed all parsed outputs, we need to request more before continuiing
```
Comment (pedroferreira1) [inline src/ui/display.c]:
```suggestion
            // We have confirmed all parsed outputs, we need to request more before continuing
```

File: src/ui/display.c [hunk 7]
```diff
@@ -259,6 +259,13 @@ bool skip_change_outputs() {
     inplace_selection_sort((size_t) G_context.tx_info.change_len, change_indices);
 
     for (uint8_t i = 0; i < G_context.tx_info.change_len; i++) {
+        // current_output is the number of parsed outputs
+        // If it is the same as confirmed_outputs we have confirmed all parsed outputs
+        if (G_context.tx_info.confirmed_outputs >= G_context.tx_info.current_output) {
```
Comment (pedroferreira1) [inline src/ui/display.c]: question(non-blocking): Why did you add this verification inside the `for` loop? It doesn't use the index, so maybe we could move it before. What do you think?

**Review by pedroferreira1 — APPROVED** [review]

File: src/ui/display.c (hunk 6 above)
Comment (r4mmer) [reply src/ui/display.c]: Changed the comment

File: src/ui/display.c (hunk 7 above)
Comment (r4mmer) [reply src/ui/display.c]: Moved the check to outside the loop.

**Review by pedroferreira1 — APPROVED** [review]
